### ISSUE TITLE
Automatically resolve NPM modules like JS does

### ISSUE THREAD

**Issue by esr360** [sass/node-sass#2445]
I think this is the most appropriate place to request this feature.

The feature was raised in the Sass repo here: https://github.com/sass/sass/issues/2520

However, it was pointed out that Sass wants to remain agnostic about package management systems as possible since Sass can be used in various languages...which is a fair comment to make.

It would be great to be able to create Sass modules which can import from node_modules without knowing its exact path, and without using a third party tool. I think every reason it exists with JS also applies to Sass these days.

Imagine how the development of JS libraries and packages would be affected if you couldn't automatically import NPM modules? Just because Sass is less popular than JS, it doesn't mean the development is any less hindered.

If I'm creating a Sass library which I will package as an NPM module, I cannot depend upon other similar NPM Sass packages without using a third party tool. Doesn't that seem mad, especially in a node.js environment?

To re-emphasize, every reason it exists with JS also applies to Sass these days.

Thanks for your consideration.

**Comment by saper** [comment]
That looks like a good case for a family of custom importers, that do roughly the same thing for node, ruby, Python, whatever...

**Comment by esr360** [comment]
@nschonni Yes, I can see that, however this issue is about providing the ability without requiring a third party tool, as explained in my post. Eyeglass doesn't provide the functionality natively which is what I'm asking, it's a third party tool.

For my own peace of mind, could you explain why this functionality won't be added natively, and why my reasons for thinking it should be added natively are incorrect?

**Comment by nschonni** [comment]
From the issue template https://raw.githubusercontent.com/sass/node-sass/master/.github/ISSUE_TEMPLATE/Feature_request.md

```
Before you request a feature, please realize this project is a wrapper for
LibSass so we can't add any language features here. If you want to add a feature
to the Sass language, please open an issue on https://github.com/sass/sass/issues/new

If you want a CLI feature, please make sure that it is supported by https://github.com/sass/sass

Thanks for making node-sass better!
```

**Comment by esr360** [comment]
@nschonni I've addressed that already - Sass won't implement natively the ability resolve NPM modules because Sass does not want to make any assumptions about the Sass compiler that is being used, nor the package manager that is being used. Node-Sass can and does make this assumption - it already knows we're in a node.js environment, so within node-sass it is a safe assumption that NPM is the package manager; hence, I believe node-sass is the most appropriate place for this feature, not Sass or Libsass. I don't believe what I'm requesting is a language feature.

For peace of mind, if you disagree, please explain why.

**Comment by saper** [comment]
@esr360 Language feature or not, I really think that a third-party module is a perfect place for a feature like this. I wish we could even split node-sass into smaller components (like https://github.com/sass/node-sass/issues/1320, https://github.com/sass/node-sass/issues/1156)

**Comment by esr360** [comment]
@saper do you disagree then that this feature is less important in a Sass ecosystem than a JavaScript one, where it is provided natively? With regards to dependency management, why is it more appropriate to use a third party module in Sass but not JavaScript?

I've provided reasons why I believe the feature should not depend upon a third party module, it would be great if someone could clarify why I'm wrong.

Frankly, I don't really care *how* we provide this functionality, but I really feel as though there should be some officially supported standard way to import from node_modules that is documented.

**Comment by saper** [comment]
I can see that you are very good at prolonged discussions. 

Let me ask a question first: what is "official support"? would you volunteer to maintain this feature, answer questions on GitHub about it, make sure it works with every new release?

**Comment by esr360** [comment]
I try to keep discussions productive.

By "official support" I mean a method that is officially backed by the creators of either Sass or Node-Sass, and included in the documentation as the "recommended way" to do it, including good examples of how to do it.

I would certainly volunteer to maintain this feature and answer questions on Github about it, and write the documentation, and contribute my free time. That's why I'm here taking the time to write this post. I enjoy participating in the open-source community.

To address your last paragraph, imagine I have developed a Sass package, `packageA`, which depends upon another Sass dependency, `packageB`, which is available on NPM. I would like my end users of packageA to simply be able to do:

```
npm install packageA
```

```scss
import '../../node_modules/packageA/packageA';
```

...without also requiring end users to configure their compiler with Eyeglass or any other third party tool.

Inside `packageA` I would naturally need to import the `packageB` dependency:

```scss
import '../../node_modules/packageB/packageB';

...
```

...which naturally won't work ultimately as the `node_modules` path is relative to `packageB`. This means that end users of `packageA` will need to either configure their compiler with Eyeglass or add their `node_modules` directory to Sass's import path, so `packageB` can be imported in `packageA` by doing:

```scss
import 'packageB/packageB';

...
```

...which is actually not too bad of a solution. If we have to tell end users of Sass NPM packages that they must manually add `node_modules` to Sass's import path, I can live with that, it just seems like a step that could easily be taken automatically be `node-sass`, and it seems like it would be a reasonable step to take.

**Comment by saper** [comment]
Does 'packageA/packageA' is mandatory to have the same 'packageA' name two times? why?

Can you review eyeglass documentation and describe how yout "Sass NPM package" would look differrently?

**Comment by esr360** [comment]
No there is no significant reason for including `packageA` twice, other than the first one to reference the directory, and the second one to reference the file. It would just be a normal path to the distribution file relative to `node_modules`.

I think there's some confusion/misunderstanding. I am approaching this issue as an author of Sass packages; requiring *users* of any Sass dependency from NPM to configure their compiler to use Eyeglass is bad Developer Experience, and the extra step may put people off using the package. I'm trying to make life easier for people who want to use NPM as a package manager for Sass packages, since that is really the only available option.

**Comment by saper** [comment]
Maybe, but what about the remaining questions?

**Comment by esr360** [comment]
Sorry which questions do you mean? I can see you asked two questions and I tried to address them both in my response. If it makes it more clear, the confusion/misunderstanding was in response to your second question; when you ask what my Sass NPM package would look like, what do you mean exactly?

The only noteworthy thing would be how it imports dependencies:

```scss
import 'path/to/dependency1';
import 'path/to/dependency2';

/* Package code... */
```

Where the paths would be relative to the *end users* `node_modules` directory, meaning `dependency1` and `dependency2` can be added as dependencies to `package.json`.

**Comment by saper** [comment]
Can you please stop repeating this Bad Developer Experience argument? I try really hard to push this discussion into something more actionable. Can you check my questions on GitHub again, since I have edited them and maybe you are not reading a full version?

**Comment by esr360** [comment]
It's a valid argument, DX is very important, and I'm sorry you don't agree.

And you are correct I had not seen your edits.

> how would you integrate this feature into node-sass? Would you modify libsass (why?) Would you modify node-sass binary binding (how)? Would you modify thin javascript wrapper? (how?)

Well, I believe node-sass is the most appropriate repository to handle this, not LibSass, like I say in my original post. I believe with Sass we can pass values to `includePaths`. If we pass a value of `node_modules` (or the user's `node_modules` directory`) automatically then this feature could be as simple to implement as that.

The node-sass documentation seems to touch on this here: https://github.com/sass/node-sass#data

**Comment by saper** [comment]
Ok, here is a little experiment (my shell is KornShell93):

    > rm -rf
    > mkdir moduleA                                               
    > echo 'var mb = require('\''moduleB'\'');' > moduleA/index.js
    > cat moduleA/index.js 
    var mb = require('moduleB');
    > node moduleA/index.js
    Module B

Should NPM Sass modules behave the same way as NPM modules? Can you explain how does the above work although there is no `node_modules` subdirectory? How would you deliver a consistent user experience to the user?

**Comment by xzyfer** [comment]
>By "official support" I mean a method that is officially backed by the creators of either Sass or Node-Sass, and included in the documentation as the "recommended way" to do it, including good examples of how to do it.

As has been stated by multiple times time, by multiple Sass engine maintainers, by multiple Sass engine repos this is not a feature that will be baked into an "official" Sass engine implementation.

I adamantly agree this position, and it's simply not up for debate at this time. 

****

However I personally have suggested there should be such an importer, and that it would be a valuable addition to the ecosystem. To that end I quickly hacked together a proof of concept. It's probably slow. It's probably buggy. But it's a start. Feel free to give it a shot a file issues.

https://www.npmjs.com/package/@node-sass/node-module-importer
